ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,308 @@
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _read_telescope_table(self, group, tel_id):
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, table=None):
+
+        if self.shower_table:
+            if (table is not None) and len(table) > 0:
+                table = join(
+                    table,
+                    self.shower_table,
+                    keys=["obs_id", "event_id"],
+                    join_type="inner"
+                )
+            else:
+                table = self.shower_table
+
+        if self.load_trigger:
+            if (table is not None) and len(table) > 0:
+                table = join(
+                    table,
+                    read_table(self.h5file, TRIGGER_TABLE),
+                    keys=["obs_id", "event_id"],
+                    join_type="inner",
+                )
+            else:
+                table = read_table(self.h5file, TRIGGER_TABLE)
+
+        if self.load_dl2_geometry:
+
+            shower_geometry_group = self.h5file.root[GEOMETRY_GROUP]
+
+            for i, reconstructor in enumerate(shower_geometry_group._v_children):
+
+                geometry = read_table(self.h5file, f"{GEOMETRY_GROUP}/{reconstructor}")
+
+                # rename DL2 columns to explicit reconstructor
+                # TBD: we could skip this if only 1 reconstructor is present
+                # or simply find another way to deal with multiple reconstructions
+                for col in set(geometry.colnames) - {"obs_id", "event_id"}:
+                    geometry.rename_column(col, f"{reconstructor}_{col}")
+
+                if ((table is None) or (len(table) == 0)) and (i == 0):
+                    table = geometry
+                else:
+                    table = join(
+                        table, geometry, keys=["obs_id", "event_id"], join_type="left"
+                    )
+        return table
+
+    def read_telescope_events(self, tel_ids):
+
+        table = vstack([self.read_telescope_events_for_id(tel_id) for tel_id in tel_ids])
+
+        return table
+
+    def read_events(self, labels=None):
+        """Read telescope-based event information.
+
+        Parameters
+        ----------
+        labels: list
+            Any list combination of tel_ids, tel_types, or telescope_descriptions.
```

Comment:
I naively tried `loader.read_events(["tel_ids"])`, assuming I would get multiple tables or something, which failed.  Better docstring is needed.

Reply:
you don't get multiple tables as you are only giving tel_ids (why a string though? it shouldn't work...)

if you use read_events_by_type you get a table for each tel_id

also the docstring says clearly 

```
Returns
        -------
        table: astropy.io.Table
            Table ...
```

as in 1 Table :)

Reply:
that's the point: I followed the docstring, and it didn't work.  I assumed a list with combination of the strings "tel_types", "tel_ids", which is of course totallly wrong, but it's not clear from the docstring.  (I basically tried using the class without any prior knowledge of how i was implemented inside, and failed)

Reply:
That may be a case to describe it more formally:
`labels: Union[List[int], List[TelescopeDescription]]`

Reply:
(also what is the difference between `tel_type` and `tel_description`? They are both representations of the same thing. A TelescopeDescription can be specified by it's class or string representation, so no need to differentiate.

Reply:
It seems a very strict interpretation of the usage to me but anyway... :)

I will explicit the types of the list elements

Reply:
> That may be a case to describe it more formally:
> `labels: Union[List[int], List[TelescopeDescription]]`

ah yes like you mentioned in the issue about type checking OK

> A TelescopeDescription can be specified by it's class or string representation, so no need to differentiate.

OK

Reply:
> `labels: Union[List[int], List[TelescopeDescription]]`

Actually it is `List[Union[int, str, TelescopeDescription]]` like in the above `get_tel_ids` function (that we should move to `SubarrayDescription`

Reply:
If we take the type hint seriously, str needs to be there, otherwise tools will complain about `read_telescope_events(['LST_LST_LSTCam'])

Reply:
Ah, it was really not clear that you could mix tel IDs and descriptions in the list.  Adding the correct type annotation helps a lot here.

Reply:
I explicited the input types in the docstrings that use the generic `label` as input